### DISCUSSION TITLE
Questions about setting up local environment for contribution

### DISCUSSION THREAD

**Discussion by ThangHuuVu** [nextauthjs/next-auth#3523] [Help] [answered]
Hi, I have some questions about settings up the local environment for contribution:
-  I can't get hot reloading for files in `src` to work anymore. I did everything mentioned on the Contributing page, with a freshly cloned repo. Is there something missing? I don't think I had this issue when I first contributed a few months ago
- For email: `npm run start:email` doesn't work. I think the correct command should be `npm run email`. Should I open a small PR for this?  https://github.com/nextauthjs/next-auth/blob/0e8be0c7d28b7304277868804cb20bd128b73d47/app/package.json#L15

**Comment by balazsorban44** [comment]
hi, I just used it yesterday. 🤔 do you have a link to your fork?

the email script should be run while inside the `app` folder

**Reply by ThangHuuVu** [reply]
Here is the link to my fork: https://github.com/ThangHuuVu/next-auth
For example, I try to change the `errors` in `src/core/pages/signin.tsx`, but it doesn't get updated.

About the email script: Here is how it look like on my side:
 <img width="209" alt="Screen Shot 2021-12-28 at 09 41 27" src="https://user-images.githubusercontent.com/31528554/147523775-40b0a431-570a-45ee-a029-122a663fd458.png">

**Reply by balazsorban44** [reply]
Ok, that might be a typo:

https://github.com/nextauthjs/next-auth/blob/0e8be0c7d28b7304277868804cb20bd128b73d47/app/package.json#L14-L15

try `npx run email`

The pages folder contains server-side rendered pages, so those won't be updated in the browser for you. you have to manually refresh. this is because we don't leverage anything from Next.js for those pages, not even in dev.

**Reply by ThangHuuVu** [reply]
> Ok, that might be a typo:
> 
> https://github.com/nextauthjs/next-auth/blob/0e8be0c7d28b7304277868804cb20bd128b73d47/app/package.json#L14-L15
> 
> try `npx run email`

Yup, that's what I propose to change in `package.json`

**Reply by ThangHuuVu** [reply]
> The pages folder contains server-side rendered pages, so those won't be updated in the browser for you. you have to manually refresh. this is because we don't leverage anything from Next.js for those pages, not even in dev.

While I can understand this, I still don't get the updated code even after refresh; which does not make sense to me. What do you think might be the reason here?

**Reply by balazsorban44** [reply]
:confused: I'm not really sure, unfortunately. But looks like you are right. Some `src` file updates seem to work with HMR, but not all of them :thinking:

I even tried upgrading the dev app dependencies: https://github.com/nextauthjs/next-auth/commit/c4936991e596e057d7536580c5b44a9b72914bba

**Reply by ThangHuuVu** [reply]
Thank you, I tried to look in the related PR https://github.com/nextauthjs/next-auth/pull/2631/files but couldn't find anything either. I'll update if I can get it work

**Comment by balazsorban44** [comment, marked as answer]
I remembered the reason for your changes not propagating in dev mode!

You have to manually remove the `app/node_modules/next-auth` folder, which gets installed with `npm run dev:setup`.

We could add it as part of the script to remove that folder.

**Reply by ThangHuuVu** [reply]
Nice catch! I found something interesting too. 
In my `node_modules`, I see that `next-auth` is at v4.0.6. It seems that `next-auth` is being installed automatically via peer dependencies of the following packages:
https://github.com/nextauthjs/next-auth/blob/c71cb8457da0febe5d71283e9361014dda652c31/app/package.json#L19-L20

I think this is the reason for changes not being reflected in dev. Because `node_modules/next-auth` v4.0.6 is found, `app` just ignore the path in `tsconfig.json` and resolve to `node_modules` instead. 

After removing these dependencies, `node_modules`, `package-lock.json` and install again:
- The changes are propagated again.
- No `next-auth` in `node_modules`, which I think makes more sense because we are using Next.js `externalDir`.

IMO, `next-auth` might need a monorepo set up to solve this dependency resolution conflict. Does this make sense?

**Reply by balazsorban44** [reply]
I've been thinking of a monorepo for `docs`, `adapters` and `core`, probably `svelte` etc. in the future (#2294).

It seems to be a big task though, as I'm kind of fond of the release process we have on this repo, and I kind of dislike using `lerna`  over at https://github.com/nextauthjs/adapters (mostly because of the lack of real experience), but I'm open to alternatives like [Changesets](https://github.com/changesets/changesets) etc.

Also, issues/discussions are scattered around multiple repos, and we did not manage to turn people to the correct repos to report certain issues. At this point, we should probably have just a single repo, and probably the standalone examples/template repo. https://github.com/nextauthjs/next-auth-example

For what it's worth, even that would be nice to have in a monorepo setup, and maybe set up some corn job that pushes any changes there. 

I also want to convert the example repo to TypeScript and drop https://github.com/nextauthjs/next-auth-typescript-example

All in all, a lot of ideas to improve maintenance! :sweat_smile:

I would be happy to see this as its own discussion and plan it properly! :pray:

**Reply by ThangHuuVu** [reply]
These are exciting improvements! My thoughts:
- One monorepo with `docs`, `adapters`, `core`, `examples`, similar to `next.js` repo. I'm surprised you did not mention `turborepo`, I think it is a promising candidate.
- TypeScript is essential, especially for new developers. However I think it's nice to have both TS and JS examples, some projects do not use TS and it could be pain for them to use the TS example. 
I am more than happy to help with anything! I do have some experiences with `monorepo` (`lerna`), but I'm open to new stuffs as well.

**Reply by ThangHuuVu** [reply]
To conclude this discussion: I still cannot dynamically load changes from `src` to `app` with the adapter dependencies; even after manual refresh. Root cause: `next-auth` is being installed in app `node_modules`.
I can think of one way to fix this: commit the `package-lock.json` file without the `next-auth` dependency in it. What do you think?

**Reply by balazsorban44** [reply]
Turborepo is something I really want to try, but currently, it does not help with versioning, which is my real pain point.

Re: your refresh issue, just do `rm -rf node_modules/next-auth`. Don't need to run `npm install` again. It did fix it for me. :thinking:
